ports/stm32/storage.c

```diff
@@ -233,41 +238,194 @@ mp_uint_t storage_write_blocks(const uint8_t *src, uint32_t block_num, uint32_t
+}
+
+STATIC mp_obj_t pyb_flash_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    // Parse arguments
+    enum { ARG_start, ARG_len };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_start, MP_ARG_INT, {.u_int = -1} },
+        { MP_QSTR_len,   MP_ARG_INT, {.u_int = -1} },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all_kw_array(n_args, n_kw, all_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+
+    if (args[ARG_start].u_int == -1 && args[ARG_len].u_int == -1) {
+        // Default singleton object that accesses entire flash, including virtual partition table
+        return MP_OBJ_FROM_PTR(&pyb_flash_obj);
+    }
+
+    pyb_flash_obj_t *self = m_new_obj(pyb_flash_obj_t);
+    self->base.type = &pyb_flash_type;
+    #if defined(SPIFLASH)
+    self->use_spi_raw = false;
+    #endif
+
+    uint32_t bl_len = (storage_get_block_count() - FLASH_PART1_START_BLOCK) * FLASH_BLOCK_SIZE;
+
+    mp_int_t start = args[ARG_start].u_int;
+    if (start == -1) {
+        start = 0;
+    } else if (0 <= start && start < bl_len && start % PYB_FLASH_ADDR_ALIGN == 0) {
+        // valid start value
+    } else {
+        mp_raise_ValueError(NULL);
+    }
 
-STATIC mp_obj_t pyb_flash_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *args) {
-    // check arguments
-    mp_arg_check_num(n_args, n_kw, 0, 0, false);
+    mp_int_t len = args[ARG_len].u_int;
+    if (len == -1) {
+        len = bl_len - start;
+    } else if (0 < len && start + len <= bl_len && len % PYB_FLASH_ADDR_ALIGN == 0) {
+        // valid len value
+    } else {
+        mp_raise_ValueError(NULL);
+    }
+
+    self->start = start;
+    self->len = len;
 
-    // return singleton object
-    return MP_OBJ_FROM_PTR(&pyb_flash_obj);
+    return MP_OBJ_FROM_PTR(self);
 }
 
-STATIC mp_obj_t pyb_flash_readblocks(mp_obj_t self, mp_obj_t block_num, mp_obj_t buf) {
+STATIC mp_obj_t pyb_flash_readblocks(size_t n_args, const mp_obj_t *args) {
+    pyb_flash_obj_t *self = MP_OBJ_TO_PTR(args[0]);
+    uint32_t block_num = mp_obj_get_int(args[1]);
     mp_buffer_info_t bufinfo;
-    mp_get_buffer_raise(buf, &bufinfo, MP_BUFFER_WRITE);
-    mp_uint_t ret = storage_read_blocks(bufinfo.buf, mp_obj_get_int(block_num), bufinfo.len / FLASH_BLOCK_SIZE);
+    mp_get_buffer_raise(args[2], &bufinfo, MP_BUFFER_WRITE);
+    mp_uint_t ret;
+    if (n_args == 3) {
+        // Cast self->start to signed in case it's pyb_flash_obj with negative start
+        block_num += FLASH_PART1_START_BLOCK + (int32_t)self->start / FLASH_BLOCK_SIZE;
+        ret = storage_read_blocks(bufinfo.buf, block_num, bufinfo.len / FLASH_BLOCK_SIZE);
+    } else {
+        uint32_t offset = mp_obj_get_int(args[3]);
+        #if MICROPY_HW_ENABLE_INTERNAL_FLASH_STORAGE
+        block_num += self->start / FLASH_BLOCK_SIZE;
+        ret = flash_bdev_readblocks_ext(bufinfo.buf, block_num, offset, bufinfo.len);
```

Comment:
`storage_read_blocks` returns unsigned, but `flash_bdev_readblocks_ext` (and `spi_bdev_readblocks_raw`) return signed.

Reply:
fixed, storage_read_blocks now returns an int